### PR TITLE
[Feature] support binary predicate for array type

### LINKED ISSUE #15365: Support comparison predicate for array type
## Feature request

`ARRAY` type in StarRocks should support comparison predicate in element-wise style, such as:
- `select [1,2,3] < [2,3,4]`
- `select [1,2,3] != [1,2,3,4]`
- `select array_sort([3,2,1]) = [1,2,3]`

NULL handling:
- `null` is considered larger than any value(be compatible with PostgreSQL)
- `select [null, 1] < [1, 1]` returns `false`
- `select [null, 1] > [1, 1]` returns `true`

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #15365

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 7 (42.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|3|7|42.86%|[261, 292, 293, 295]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 7 (42.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|3|7|42.86%|[261, 292, 293, 295]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: be/src/exprs/vectorized/binary_predicate.cpp [hunk 1]
```diff
@@ -48,6 +49,33 @@ using EvalGt = std::greater<typename PredicateCmpType<ptype>::CmpType>;
 template <LogicalType ptype>
 using EvalGe = std::greater_equal<typename PredicateCmpType<ptype>::CmpType>;
 
+struct EvalCmp {
```
Comment (satanson) [inline be/src/exprs/vectorized/binary_predicate.cpp:52]: here only compared the cmp_value to zero, can wo rename EvalCmp to EvalZeroCmp?

File: be/src/exprs/vectorized/binary_predicate.cpp [hunk 2]
```diff
@@ -72,6 +100,35 @@ class VectorizedBinaryPredicate final : public Predicate {
...
+        auto lhs_arr = std::static_pointer_cast<ArrayColumn>(l);
+        auto rhs_arr = std::static_pointer_cast<ArrayColumn>(r);
+
+        ColumnBuilder<TYPE_BOOLEAN> builder(ptr->num_rows());
+        std::vector<int8_t> cmp_result;
+        lhs_arr->compare_column(*rhs_arr, &cmp_result);
+
+        // Convert the compare result (-1, 0, 1) to the predicate result (true/false)
+        for (int8_t cmp : cmp_result) {
+            bool pred_res = _comparator(cmp);
```
Comment (satanson) [inline be/src/exprs/vectorized/binary_predicate.cpp:122]: do you dissamble this code to ensure that simd instruction sequence can be generated from this for-loop?
another thing is that spliting compare proceduce into two procedures cost less time than one compare procedure?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/binary_predicate.cpp]: fixed

**Comment by mofeiatwork** [comment]
run starrocks_fe_unittest

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: be/src/column/array_column.cpp [hunk 3]
```diff
@@ -382,6 +382,16 @@ int ArrayColumn::compare_at(size_t left, size_t right, const Column& right_colum
     return lhs_size < rhs_size ? -1 : (lhs_size == rhs_size ? 0 : 1);
 }
 
+void ArrayColumn::compare_column(const Column& rhs_column, std::vector<int8_t>* output) const {
+    CHECK(size() == rhs_column.size()) << "Two input columns must have same rows";
```
Comment (silverbullet233) [inline be/src/column/array_column.cpp:386]: why we have this limitation?what will happen if we compare two array with different length?

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 7 (42.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/BinaryPredicate.java|3|7|42.86%|[261, 292, 293, 295]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**